aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreError.java

```diff
@@ -32,37 +33,59 @@
  */
 public final class DataStoreError<M extends Model> {
     private final GraphQLResponse.Error error;
+    private final Operation operation;
     private final M local;
     private final M remote;
 
     /**
      * Constructs a new error instance containing details regarding
      * a failed publication to the cloud and involved models.
-     * @param error the error response from the server
-     * @param local the local model pending mutation
-     * @param remote the remote model if applicable
+     *
+     * @param error     the error response from the server
+     * @param operation the type of operation
+     * @param local     the local model pending mutation
+     * @param remote    the remote model if applicable
      */
     public DataStoreError(
             @NonNull GraphQLResponse.Error error,
+            @NonNull Operation operation,
```

Comment:
If the `DataStoreError` is packaged under `com.amplifyframework.datastore`, then we should try to treat these two fields as implementation concerns (or, in other words, do not provide getters for them.)

Right now, the `DataStoreError` sort of looks like an API thing by a different name. You could also call this an `AppSyncError`, or a `MutationError`, or something like that?

Reply:
Considering how similar the error handler is to the conflict handler in the same package, I figured they should be grouped together. I considered making this error class an internal class inside `DataStoreErrorHandler` like `ConflictData` is inside `DataStoreConflictHandler`, but I left it as a standalone class for arbitrary reasons.

As for the not exposing `GraphQLResponse.Error` and `Operation`, what is the purpose? Wouldn't they be the most important bits of information that the customer might want to know when handling the error? Besides, I am trying to reach functional parity with JS here by exposing all of these properties.

Lastly, I decided to leave the name more generic (I did initially name it `MutationError` like you suggested) in case that we might want to support generic error handling for any GraphQL error in the future. Also, I invoke the error handler for any GraphQL error from mutation publication at the moment (not sure if JS does for only AppSync-related error or not), so `AppSyncError` would be less appropriate.